internal/wasi/fs.go

```diff
@@ -111,17 +148,159 @@ func (f *memFile) Seek(offset int64, whence int) (int64, error) {
 	case io.SeekCurrent:
 		f.offset += offset
 	case io.SeekEnd:
-		f.offset = int64(len(f.buf)) + offset
+		f.offset = int64(len(f.fsEntry.Contents)) + offset
 	default:
 		return 0, fmt.Errorf("invalid whence: %d", whence)
 	}
 	return f.offset, nil
 }
 
-// Close implements io.Closer
-func (f *memFile) Close() error {
-	if f.flush != nil {
-		f.flush(f.buf)
+// memDir represents an opened memFS directory.
+// memDir implements fs.File, fs.ReadDirFile, and io.Seeker.
+type memDir struct {
+	// name of the opened file, not a full path but a base name.
+	name string
+	// Seek offset of this opened directory.
+	offset int64
+	// Sorted file list of the dirEntries for ReadDir and Seek.
+	entries []fs.DirEntry
+	// The memFSEntry this memFileInfo is about.
+	fsEntry *memFSEntry
+}
+
+func newMemDir(dir *memFSEntry, baseName string) *memDir {
+	// Cache the file list of the directory for ReadDir to return the result in the consistent order.
```

Comment:
caching a read could be a performance problem. If only caching for tests, maybe sort in the test instead?

Reply:
> If only caching for tests

cache is not for tests, just to avoid sorting the file list every time ReadDir is called. (btw, [sort is required for ReadDir](https://github.com/tetratelabs/wazero/pull/362/files#r828715370)) If your intention is caching it is weird or too much, yes we can sort the list inside ReadDir instead of here. wdyt? thanks!

Reply:
yeah I think it is better to keep the abstraction coherent. fs.FS shouldn't bleed too much into how we implement WASI, that way in case it isn't used, there's no perf hit. In the call site that requires caching, you can add a comment that it is required by the definition of fs.FS

Reply:
accidentally added a comment and clicked resolve. What I meant was add like you said sorting to ReadDir. Don't cache on first impl as it is possible no-one calls readDir and each thing we add to this PR == more code to review and edge case unit tests